http/vibe/http/auth/digest_auth.d

```diff
@@ -179,3 +180,132 @@ string createDigestPassword(string realm, string user, string password)
 }
 
 alias DigestHashCallback = string delegate(string realm, string user);
+
+/// Structure which describes requirements of the digest authentication - see https://tools.ietf.org/html/rfc2617
+struct DigestAuthParams {
+	enum Qop { none = 0, auth = 1, auth_int = 2 }
+	enum Algorithm { none = 0, md5 = 1, md5_sess = 2 }
+
+	string realm, domain, nonce, opaque;
+	Algorithm algorithm = Algorithm.md5;
+	bool stale;
+	Qop qop;
+
+	/// Parses WWW-Authenticate header value with the digest parameters
+	this(string auth) {
+		import std.algorithm : splitter;
+
+		assert(auth.startsWith("Digest "), "Correct Digest authentication request not provided");
+
+		foreach (param; auth["Digest ".length..$].splitter(','))
+		{
+			auto idx = param.indexOf("=");
+			if (idx <= 0) {
+				logError("Invalid parameter in auth header: %s (%s)", param, auth);
+				continue;
+			}
+			auto k = param[0..idx];
+			auto v = param[idx+1..$];
+			switch (k.strip().toLower()) {
+				default: break;
+				case "realm": realm = v[1..$-1]; break;
+				case "domain": domain = v[1..$-1]; break;
+				case "nonce": nonce = v[1..$-1]; break;
+				case "opaque": opaque = v[1..$-1]; break;
+				case "stale": stale = v.toLower() == "true"; break;
+				case "algorithm":
+					switch (v) {
+						default: break;
+						case "MD5": algorithm = Algorithm.md5; break;
+						case "MD5-sess": algorithm = Algorithm.md5_sess; break;
+					}
+					break;
+				case "qop":
+					foreach (q; v[1..$-1].splitter(',')) {
+						switch (q) {
+							default: break;
+							case "auth": qop |= Qop.auth; break;
+							case "auth-int": qop |= Qop.auth_int; break;
+						}
+					}
+					break;
+			}
+		}
+	}
+}
+
+/**
+	Creates the digest authorization request header.
+
+	Params:
+		username = user name
+		password = user password
+		url = requested url
+		auth = value from the WWW-Authenticate response header
+		cnonce = client generated unique data string (required only when some qop is requested)
+		nc = the count of requests sent by the client (required only when some qop is requested)
+		entityBody = request entity body required only if qop==auth-int
+*/
+auto createDigestAuthHeader(U)(HTTPMethod method, U url, string username, string password, DigestAuthParams auth,
```

Comment:
I'd split this up into two methods:
```
void addDigestAuth(scope HTTPRequest req, ...);  // analogous to addBasicAuth()
void writeDigestAuthHeader(R)(ref R dst, ...); // dst is an output range that receives the contents
```
The latter can probably be left private for now (to keep the API small and avoid the need for deprecation processes if something ever needs to change).